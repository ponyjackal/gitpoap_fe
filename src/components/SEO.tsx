import Head from 'next/head';

type SEOProps = {
  title: string;
  description: string;
  image: string;
  url: string;
};

export const SEO = ({ title, description, image, url }: SEOProps) => {
  return (
    <Head>
      <title>{title}</title>
      <meta key="description" name="description" content={description} />

      {/* <!-- Metadata for link cards --> */}
      <meta key="og:site_name" property="og:site_name" content="GitPOAP " />
      <meta key="og:title" property="og:title" content={title} />
      <meta key="og:type" property="og:type" content="website" />
      <meta key="og:url" property="og:url" content={url} />
      <meta key="og:image" property="og:image" content={image} />
      <meta key="og:image:type" property="og:image:type" content="image/png" />
      <meta key="og:image:width" property="og:image:width" content="256" />
      <meta key="og:image:height" property="og:image:height" content="256" />
      <meta key="og:description" property="og:description" content={description} />

      <meta key="twitter:card" property="twitter:card" content="summary" />
      <meta key="twitter:url" property="twitter:url" content={url} />
      <meta key="twitter:title" property="twitter:title" content={title} />
      <meta key="twitter:description" property="twitter:description" content={description} />
      <meta key="twitter:image" property="twitter:image" content={image} />
    </Head>
  );
};
