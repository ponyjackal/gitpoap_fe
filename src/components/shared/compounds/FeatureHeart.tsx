import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Loader } from '@mantine/core';
import { FaHeart } from 'react-icons/fa';
import { ExtraHover, ExtraPressed, TextAccent, TextLight } from '../../../colors';
import { useFeaturedPOAPs, useFeaturedPOAPsDispatch } from '../../profile/FeaturedPOAPsContext';

type Props = {
  className?: string;
  poapTokenId: string;
};

const HeartWrapper = styled.div<{ isFeatured: boolean }>`
  > svg {
    fill: ${({ isFeatured }) => (isFeatured ? TextAccent : TextLight)};
  }
`;

const FeatureHeartStyled = styled(FaHeart)`
  color: ${TextLight};
  height: ${rem(18)};
  width: ${rem(18)};
  transition: 150ms fill ease;
  fill: ${TextLight};

  &:hover {
    fill: ${ExtraHover};
    cursor: pointer;
  }
  &:active {
    fill: ${ExtraPressed};
  }
`;

export const FeatureHeart = ({ className, poapTokenId }: Props) => {
  const {
    featuredPOAPsState: { featuredPOAPTokenIDs },
    showHearts,
    loadingIds,
  } = useFeaturedPOAPs();
  const { addFeaturedPOAP, removeFeaturedPOAP } = useFeaturedPOAPsDispatch();
  const isFeatured: boolean = !!featuredPOAPTokenIDs && !!featuredPOAPTokenIDs[poapTokenId];
  const isFeaturedLoading = !!loadingIds && !!loadingIds[poapTokenId];

  if (!showHearts) {
    return null;
  }

  return (
    <HeartWrapper isFeatured={isFeatured}>
      {isFeaturedLoading ? (
        <Loader className={className} size={18} />
      ) : (
        <FeatureHeartStyled
          className={className}
          onClick={() => {
            if (isFeatured) {
              removeFeaturedPOAP(poapTokenId);
            } else {
              addFeaturedPOAP(poapTokenId);
            }
          }}
        />
      )}
    </HeartWrapper>
  );
};
