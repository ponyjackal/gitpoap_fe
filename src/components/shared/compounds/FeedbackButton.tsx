import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Popover, Box } from '@mantine/core';
import { useDisclosure } from '@mantine/hooks';
import { GitPOAPLogoNoText } from '../elements/icons';
import { BackgroundPanel, BackgroundPanel2, ExtraHover, ExtraPressed } from '../../../colors';
import { Text } from '../elements';

const StyledLogo = styled(GitPOAPLogoNoText)`
  height: ${rem(45)};
  width: ${rem(45)};
  transition: transform 200ms ease-in-out;
  path {
    transition: fill 200ms ease-in-out;
  }
`;

const CircleButton = styled.a`
  display: flex;
  align-items: center;
  justify-content: center;
  height: ${rem(70)};
  width: ${rem(70)};
  background-color: ${BackgroundPanel};
  transition: background-color 200ms ease-in-out, transform 200ms ease-in-out;
  border-radius: 50%;
  border: ${rem(3)} solid ${BackgroundPanel2};
  &:hover {
    background-color: ${BackgroundPanel2};
    transform: scale(1.1);
    cursor: pointer;
    ${StyledLogo} {
      transform: scale(1.1);
      path {
        fill: ${ExtraHover};
      }
    }
  }
  &:active {
    background-color: ${BackgroundPanel};
    ${StyledLogo} {
      color: ${ExtraHover};
      path {
        fill: ${ExtraPressed};
      }
    }
  }
`;

const PopoverContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

type Props = {
  className?: string;
  href: string;
};

export const FeedbackButton = ({ className, href }: Props) => {
  const [isOpen, { open, close }] = useDisclosure(false);
  return (
    <Popover
      opened={isOpen}
      onClose={close}
      position="top"
      withArrow
      trapFocus={false}
      closeOnEscape={false}
      transition="fade"
      transitionDuration={200}
      radius="lg"
      withinPortal
      zIndex={100}
      styles={{
        dropdown: {
          backgroundColor: BackgroundPanel,
        },
      }}
    >
      <Popover.Target>
        <Box onMouseEnter={open} onMouseLeave={close} className={className} style={{ zIndex: 100 }}>
          <CircleButton href={href} target="_blank" rel="noopener noreferrer">
            <StyledLogo />
          </CircleButton>
        </Box>
      </Popover.Target>
      <Popover.Dropdown>
        <PopoverContainer>
          <Text>{'Give Feedback'}</Text>
        </PopoverContainer>
      </Popover.Dropdown>
    </Popover>
  );
};
