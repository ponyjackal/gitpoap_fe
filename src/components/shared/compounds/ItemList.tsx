import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Header } from '../elements/Header';
import { Button } from '../elements/Button';
import { FaPlus } from 'react-icons/fa';
import { Select } from '../elements/Select';
import { Text } from '../elements/Text';
import { TextGray } from '../../../colors';
import { Input } from '../elements/Input';

type Props = {
  title?: string;
  className?: string;
  children?: React.ReactNode;
  selectOptions: SelectOption[];
  selectValue: string;
  onSelectChange: (value: string) => void;
  isLoading: boolean;
  hasShowMoreButton: boolean;
  showMoreOnClick: () => void;
  searchInputPlaceholder?: string;
  searchInputValue?: string;
  onSearchInputChange?: React.ChangeEventHandler<HTMLInputElement>;
};

export type SelectOption<T = string> = { value: T; label: string };

const Container = styled.div`
  display: inline-flex;
  flex-direction: column;
  width: 100%;
`;

const Heading = styled.div`
  display: inline-flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const ListTitle = styled(Header)`
  font-size: ${rem(30)};
  line-height: ${rem(42)};
`;

const Sorting = styled.div`
  display: inline-flex;
  flex-direction: row;
`;

const ShowMore = styled(Button)`
  align-self: center;
`;

const SortBy = styled(Text)`
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: ${rem(12)};
  line-height: ${rem(18)};
  letter-spacing: ${rem(2)};
  text-transform: uppercase;
  color: ${TextGray};
  margin-right: ${rem(10)};
`;

const SearchInput = styled(Input)`
  margin-right: ${rem(50)};
`;

export const ItemList = ({
  children,
  className,
  title,
  selectOptions,
  selectValue,
  onSelectChange,
  isLoading,
  hasShowMoreButton,
  showMoreOnClick,
  searchInputPlaceholder,
  searchInputValue,
  onSearchInputChange,
}: Props) => {
  return (
    <Container className={className}>
      <Heading>
        <ListTitle>{title}</ListTitle>
        <Sorting>
          {searchInputValue !== undefined && onSearchInputChange && (
            <SearchInput
              placeholder={searchInputPlaceholder}
              value={searchInputValue}
              onChange={onSearchInputChange}
            />
          )}
          <SortBy>{'Sort By: '}</SortBy>
          <Select data={selectOptions} value={selectValue} onChange={onSelectChange} />
        </Sorting>
      </Heading>
      {children}
      {hasShowMoreButton && (
        <ShowMore
          onClick={showMoreOnClick}
          leftIcon={<FaPlus />}
          variant="outline"
          loading={isLoading}
        >
          {'Show more'}
        </ShowMore>
      )}
    </Container>
  );
};
