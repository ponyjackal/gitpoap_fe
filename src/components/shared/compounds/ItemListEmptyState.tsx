import React from 'react';
import { rem } from 'polished';
import { Group } from '@mantine/core';

type Props = {
  icon?: React.ReactNode;
  children: React.ReactNode;
};

export const EmptyState = ({ children, icon }: Props) => {
  return (
    <Group
      direction="column"
      position="center"
      align="center"
      grow
      style={{ padding: rem(60), flex: '1', gridColumnEnd: 1, gridRowEnd: 2 }}
    >
      {icon && icon}
      {children}
    </Group>
  );
};
