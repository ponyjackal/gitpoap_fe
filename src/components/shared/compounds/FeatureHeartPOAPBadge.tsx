import React from 'react';
import { Loader } from '../elements/Loader';
import { useFeaturedPOAPsDispatch } from '../../profile/FeaturedPOAPsContext';
import { HeartWrapper, FeatureHeartStyled } from './FeatureHeart';

type Props = {
  className?: string;
  poapTokenId: string;
  isFeatured: boolean;
  isFeaturedLoading: boolean;
};

export const FeaturedHeartPOAPBadge = ({
  className,
  poapTokenId,
  isFeatured,
  isFeaturedLoading,
}: Props) => {
  const { addFeaturedPOAP, removeFeaturedPOAP } = useFeaturedPOAPsDispatch();
  return (
    <HeartWrapper isFeatured={isFeatured}>
      {isFeaturedLoading ? (
        <Loader className={className} size={18} />
      ) : (
        <FeatureHeartStyled
          className={className}
          onClick={() => {
            if (isFeatured) {
              removeFeaturedPOAP(poapTokenId);
            } else {
              addFeaturedPOAP(poapTokenId);
            }
          }}
        />
      )}
    </HeartWrapper>
  );
};
