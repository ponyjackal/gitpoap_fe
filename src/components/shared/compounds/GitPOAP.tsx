import { Link } from '../../Link';
import React from 'react';
import styled, { css } from 'styled-components';
import { rem } from 'polished';
import { GitPOAPBadge } from '../elements/GitPOAPBadge';
import { Title } from '../elements/Title';
import { TextLight } from '../../../colors';
import { FeatureHeart } from './FeatureHeart';

type Props = {
  gitPOAPId: number;
  imgSrc: string;
  name: string;
  orgName: string;
  description?: string;
  className?: string;
  poapTokenId?: string;
  onClick?: () => void;
};

const LineClamp = (lines: number) => css`
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: ${lines};
  -webkit-box-orient: vertical;
`;

const Wrapper = styled.div`
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Info = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  max-width: ${rem(150)};
`;

const TitleStyled = styled(Title)`
  margin-top: ${rem(10)};
  text-align: center;
  ${LineClamp(3)};
`;

const OrgName = styled.div`
  font-family: PT Mono;
  font-style: normal;
  font-weight: bold;
  font-size: ${rem(11)};
  line-height: ${rem(18)};
  text-align: center;
  letter-spacing: ${rem(1.2)};
  text-transform: uppercase;
  color: ${TextLight};
  margin-top: ${rem(8)};
`;

const Description = styled.div`
  font-family: PT Mono;
  font-style: normal;
  font-weight: normal;
  font-size: ${rem(11)};
  line-height: ${rem(14)};
  text-align: center;
  letter-spacing: ${rem(-0.1)};
  color: ${TextLight};
  margin-top: ${rem(8)};
  ${LineClamp(2)};
`;

const Heart = styled(FeatureHeart)`
  position: absolute;
  bottom: ${rem(0)};
  right: ${rem(10)};
`;

const BadgeWrapper = styled(Wrapper)`
  position: relative;
`;

export const GitPOAP = ({
  className,
  poapTokenId,
  gitPOAPId,
  imgSrc,
  name,
  orgName,
  description,
  onClick,
}: Props) => {
  return (
    <Wrapper className={className}>
      <BadgeWrapper>
        <Link href={`/gp/${gitPOAPId}`} passHref>
          <GitPOAPBadge
            size="sm"
            imgUrl={imgSrc}
            onClick={() => {
              if (onClick) {
                onClick();
              }
            }}
          />
          {poapTokenId && <Heart poapTokenId={poapTokenId} />}
        </Link>
      </BadgeWrapper>
      <Info>
        <Link href={`/gp/${gitPOAPId}`} passHref>
          <TitleStyled>{name}</TitleStyled>
        </Link>
        <OrgName>{orgName}</OrgName>
        {description && <Description>{description}</Description>}
      </Info>
    </Wrapper>
  );
};
