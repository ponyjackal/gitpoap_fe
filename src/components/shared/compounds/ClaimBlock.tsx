import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Button } from '../elements/Button';
import { GitPOAP } from './GitPOAP';
import { FaCheckCircle, FaCoins, FaEthereum } from 'react-icons/fa';

type Props = {
  gitPOAPId: number;
  imgSrc: string;
  name: string;
  orgName: string;
  description: string;
  onClickClaim: () => void;
  onClickBadge?: () => void;
  isClaimed?: boolean;
  isLoading?: boolean;
  isConnected?: boolean;
};

const Wrapper = styled.div`
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  height: ${rem(340)};
`;

const ButtonWrapper = styled.div`
  flex: 1;
  margin-top: ${rem(14)};
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
`;

const getButtonText = (isClaimed: boolean | undefined, isConnected: boolean | undefined) => {
  if (isClaimed) {
    return 'Minted';
  }

  if (!isConnected) {
    return 'Connect';
  }

  return 'Mint';
};

export const ClaimBlock = ({
  gitPOAPId,
  imgSrc,
  name,
  orgName,
  description,
  onClickClaim,
  onClickBadge,
  isClaimed,
  isLoading,
  isConnected,
}: Props) => {
  return (
    <Wrapper>
      <GitPOAP
        gitPOAPId={gitPOAPId}
        imgSrc={imgSrc}
        name={name}
        orgName={orgName}
        description={description}
        onClick={onClickBadge}
      />
      <ButtonWrapper>
        <Button
          onClick={onClickClaim}
          loading={isLoading}
          leftIcon={isClaimed ? <FaCheckCircle /> : !isConnected ? <FaEthereum /> : <FaCoins />}
          disabled={isClaimed}
        >
          {getButtonText(isClaimed, isConnected)}
        </Button>
      </ButtonWrapper>
    </Wrapper>
  );
};
