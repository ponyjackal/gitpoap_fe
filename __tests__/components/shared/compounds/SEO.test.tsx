import { render } from '@testing-library/react';
import 'jest-styled-components';
import { SEO } from '../../../../src/components/shared/compounds/SEO';

jest.mock('next/head', () => {
  return {
    __esModule: true,
    default: ({ children }: { children: Array<React.ReactElement> }) => {
      return <>{children}</>;
    },
  };
});

const renderSEO = () => {
  return render(
    <SEO
      title={'test-title'}
      description={'test-description'}
      image={'/test/test.png'}
      url={'/test/test-link'}
    />,
    {
      container: document.head,
    },
  );
};

describe('SEO', () => {
  it('should render with title, description, image and url', () => {
    const { container, baseElement } = renderSEO();
    const seo = baseElement.parentElement?.firstChild as HTMLHeadElement;
    expect(seo).toMatchSnapshot();

    // const title = screen.getByText('test-title')

    const title = seo.querySelector('title');
    expect(title).toBeInTheDocument();
    expect(title).toMatchSnapshot();

    const metaDescription = seo.querySelector('meta[property="og:description"]');
    expect(metaDescription).toBeInTheDocument();
    expect(metaDescription).toHaveAttribute('content', 'test-description');
    expect(metaDescription).toMatchSnapshot();

    const metaTitle = seo.querySelector('meta[property="og:title"]');
    expect(metaTitle).toBeInTheDocument();
    expect(metaTitle).toHaveAttribute('content', 'test-title');
    expect(metaTitle).toMatchSnapshot();

    const metaUrl = seo.querySelector('meta[property="og:url"]');
    expect(metaUrl).toBeInTheDocument();
    expect(metaUrl).toHaveAttribute('content', '/test/test-link');
    expect(metaUrl).toMatchSnapshot();

    const metaImage = seo.querySelector('meta[property="og:image"]');
    expect(metaImage).toBeInTheDocument();
    expect(metaImage).toHaveAttribute('content', '/test/test.png');
    expect(metaImage).toMatchSnapshot();
  });
});
