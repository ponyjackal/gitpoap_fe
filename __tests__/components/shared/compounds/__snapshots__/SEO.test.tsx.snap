// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO should render with title, description, image and url 1`] = `
<head>
  <title>
    test-title
  </title>
  <meta
    content="test-description"
    name="description"
  />
  <meta
    content="GitPOAP "
    property="og:site_name"
  />
  <meta
    content="test-title"
    property="og:title"
  />
  <meta
    content="website"
    property="og:type"
  />
  <meta
    content="/test/test-link"
    property="og:url"
  />
  <meta
    content="/test/test.png"
    property="og:image"
  />
  <meta
    content="image/png"
    property="og:image:type"
  />
  <meta
    content="256"
    property="og:image:width"
  />
  <meta
    content="256"
    property="og:image:height"
  />
  <meta
    content="test-description"
    property="og:description"
  />
  <meta
    content="summary"
    property="twitter:card"
  />
  <meta
    content="/test/test-link"
    property="twitter:url"
  />
  <meta
    content="test-title"
    property="twitter:title"
  />
  <meta
    content="test-description"
    property="twitter:description"
  />
  <meta
    content="/test/test.png"
    property="twitter:image"
  />
  <style
    data-styled="active"
    data-styled-version="5.3.5"
  />
</head>
`;

exports[`SEO should render with title, description, image and url 2`] = `
<title>
  test-title
</title>
`;

exports[`SEO should render with title, description, image and url 3`] = `
<meta
  content="test-description"
  property="og:description"
/>
`;

exports[`SEO should render with title, description, image and url 4`] = `
<meta
  content="test-title"
  property="og:title"
/>
`;

exports[`SEO should render with title, description, image and url 5`] = `
<meta
  content="/test/test-link"
  property="og:url"
/>
`;

exports[`SEO should render with title, description, image and url 6`] = `
<meta
  content="/test/test.png"
  property="og:image"
/>
`;
